### PR TITLE
api: LEDS api introduction

### PR DESCRIPTION
**Pull request by adisuissa**
Commit Message: Adding the LEDS api
Additional Description:
This is the first part of the work on #10373.
Here we introduce the API changes in the load assignments configurations.
LEDS is only supported in delta-xDS (incremental).

Risk Level: Low - api changes only.
Testing: None - api changes only.
Docs Changes: None at the moment.
Release Notes: None at the moment.
Platform Specific Features: N/A.

Signed-off-by: Adi Suissa-Peleg <adip@google.com>

### CODE REVIEW

File: api/envoy/config/endpoint/v3/endpoint_components.proto [hunk 1]
```diff
@@ -121,7 +134,14 @@ message LocalityLbEndpoints {
   core.v3.Locality locality = 1;
 
   // The group of endpoints belonging to the locality specified.
-  repeated LbEndpoint lb_endpoints = 2;
+  repeated LbEndpoint lb_endpoints = 2
+      [(udpa.annotations.field_migrate).oneof_promotion = "lb_config"];
```
Comment (markdroth) [inline api/envoy/config/endpoint/v3/endpoint_components.proto:138]: I don't think it's legal to have a repeated field in a oneof, so this oneof_promotion won't actually work. :(

Comment (adisuissa) [reply api/envoy/config/endpoint/v3/endpoint_components.proto:138]: Yes you are correct, thanks!
I think the right way would be to support both `lb_endpoints` and the `leds_cluster_locality_config` (similar to what's being done for VHDS). I've looked at my implementation and it won't require much change.

Comment (markdroth) [reply api/envoy/config/endpoint/v3/endpoint_components.proto]: Alternatively, consider deprecating the existing `lb_endpoints` field and replacing it with a message that contains the repeated field, and then putting *that* message in the oneof.  In other words, add this message:

```
message LbEndpoints {
  repeated LbEndpoint lb_endpoints = 1;
}
```

And then change the fields here to:

```
// Deprecated -- use inline_endpoints instead.
repeated LbEndpoint lb_endpoints = 2 [deprecated = true];

oneof endpoints {
  LbEndpoints inline_endpoints = 7;
  LedsClusterLocalityConfig leds_cluster_locality_config = 8;
}
```

**Review by markdroth — COMMENTED** [review]
/wait

**Comment by adisuissa** [comment]
Thanks for the suggestion.
I've considered this, and I'm not sure if the deprecation churn is worth the price (although I might be wrong, and would prefer to hear this now :) ).
On the one hand, a single source of truth where all the endpoints are located, either as a list in an EDS response or a list coming from LEDS, makes sense to me from operational point-of-view. On the other hand, to reduce deprecation churn and to align with VHDS allowing both fields concurrently is probably the right way to go.
Feel free to tell me if you think differently.

**Comment by markdroth** [comment]
I think either way would work.

I don't find the argument of consistency with VHDS that compelling.  I think that functionality was added to VHDS because there was a use-case for it, but it's not clear to me that we have a corresponding use-case for LEDS.  But if there is such a use-case, then I think that would be a strong argument.

I tend to favor clarity of the API over deprecation pain, but I do see that pain as well.  Others may feel differently.

I'd be interested to hear what others think.

**Comment by adisuissa** [comment]
/cc @envoyproxy/api-shepherds 
Currently in a Locality there's a field:
```
repeated LbEndpoint lb_endpoints = 2;
```

With the introduction of LEDS we can do one of the following:
1. Deprecate the field, and add a `oneof` around the LEDS config and a new message that has the repeated LbEndpoint (See [comment above](https://github.com/envoyproxy/envoy/pull/17419#discussion_r674241484)). 
Pros: endpoints for a locality are configured either via LEDS or in the EDS response (single source of truth), which IMHO seems easier to control from an operations point-of-view.
Cons: churn due to field deprecation.
2. Keep `lb_endpoints`, and add the LEDS config field (See [current change](https://github.com/envoyproxy/envoy/pull/17419/files#diff-7f4d2f3972135f5812ccbbf85dd035806237f31eb8612a09aea146ce581840fb)) so the locality's endpoints will essentially be merged from the 2 sources.
Pros: This is consistent with VHDS (and the `virtual_hosts` field), and for Envoy the code can easily support this.
Cons: If there isn't a use-case for this, then the API shouldn't allow it.

Any thoughts on this?

**Comment by htuch** [comment]
I think we could just require that they are mutually exclusive in documentation and enforce in code rather than proto structure.

**Comment by adisuissa** [comment]
Deprecating in the proto structure is probably the right way to go (from API standpoint).
I've left a TODO to deprecate the `lb_endpoints` list and replace it as @markdroth suggested. This will be done once LEDS is operational.

**Comment by adisuissa** [comment]
@lizan @mattklein123 for non-googler review. PTAL

**Review by lizan — APPROVED** [review]

**Review by mattklein123 — APPROVED** [review]
Thanks, LGTM.

**Pull request merged**